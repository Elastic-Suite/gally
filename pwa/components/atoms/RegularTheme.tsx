import { createTheme } from '@mui/material/styles'
import { defaultTheme } from 'react-admin'

declare module '@mui/material/styles/createPalette' {
  interface Palette {
    neutral: Palette['primary']
    menu: {
      text500
      text600
      hover
      active
    }
  }
  interface PaletteOptions {
    neutral: PaletteOptions['primary']
    menu: {
      text500
      text600
      hover
      active
    }
  }
}

/* Creation of custom light theme
 * There is always scss assets to scss modifications but here is to be used by Material UI
 * see : https://mui.com/customization/theming/
 */
const regularTheme = createTheme({
  ...defaultTheme,
  palette: {
    primary: {
      light: '#FBC0B9',
      main: '#ED7465',
      dark: '#A02213',
    },
    secondary: {
      light: '#BABDFC',
      main: '#2C19CD',
      dark: '#1812A0',
    },
    neutral: {
      light: '#F4F7FF',
      main: '#B5B9D9',
      dark: '#2F3674',
    },
    error: {
      light: '#FFE7E4',
      main: '#A02213',
    },
    warning: {
      light: '#FEF9D0',
      main: '#60590D',
    },
    success: {
      light: '#E7F4EC',
      main: '#18753C',
    },
    menu: {
      text500: '#424880',
      text600: '#212250',
      hover: '#E7E8FF',
      active: '#2C19CD',
    },
    mode: 'light',
  },
  typography: {
    fontFamily: 'Inter',
    h1: {
      fontSize: 36,
      lineHeight: '44px',
      fontWeight: 600,
    },
    h2: {
      fontSize: 28,
      lineHeight: '38px',
      fontWeight: 600,
    },
    h3: {
      fontSize: 24,
      lineHeight: '32px',
      fontWeight: 600,
    },
    h4: {
      fontSize: 20,
      lineHeight: '30px',
      fontWeight: 600,
    },
    h5: {
      fontSize: 18,
      lineHeight: '28px',
      fontWeight: 600,
    },
    h6: {
      fontSize: 16,
      lineHeight: '24px',
      fontWeight: 600,
    },
    body1: {
      fontSize: 14,
      lineHeight: '20px',
      fontWeight: 500,
    },
    body2: {
      fontSize: 14,
      lineHeight: '20px',
      fontWeight: 400,
    },
    caption: {
      fontSize: 12,
      lineHeight: '18px',
      fontWeight: 400,
    },
  },
  components: {
    MuiCollapse: {
      styleOverrides: {
        wrapper: {
          width: 'inherit',
        },
        wrapperInner: {
          width: 'inherit',
        },
      },
    },
  },
})

export default regularTheme
