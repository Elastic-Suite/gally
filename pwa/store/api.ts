import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { fetchApi } from '~/services'
import { IDocsJson, IDocsJsonld, IFetch, LoadStatus } from '~/types'

import { IThunkApi, RootState } from './store'

export interface IDocs {
  json: IDocsJson
  jsonld: IDocsJsonld
}

export interface IApiState {
  docs: IFetch<IDocs>
}

const initialState: IApiState = {
  docs: {
    data: {
      json: null,
      jsonld: null,
    },
    error: null,
    status: LoadStatus.IDLE,
  },
}

export const fetchDocs = createAsyncThunk<IDocs, void, IThunkApi>(
  'api/docs.jsonld',
  async (_, { getState }) => {
    const {
      i18n: { language },
    } = getState()
    const results = await Promise.all([
      fetchApi<IDocsJson>(language, '/docs.json'),
      fetchApi<IDocsJsonld>(language, '/docs.jsonld'),
    ])
    return { json: results[0], jsonld: results[1] }
  }
)

const apiSlice = createSlice({
  name: 'api',
  initialState,
  reducers: {},
  extraReducers(builder) {
    builder
      .addCase(fetchDocs.pending, (state) => {
        state.docs.status = LoadStatus.LOADING
      })
      .addCase(fetchDocs.fulfilled, (state, action) => {
        state.docs.status = LoadStatus.SUCCEEDED
        state.docs.data = action.payload
      })
      .addCase(fetchDocs.rejected, (state, action) => {
        state.docs.status = LoadStatus.FAILED
        state.docs.error = action.error.message
      })
  },
})

// export const { } = apiSlice.actions
export const apiReducer = apiSlice.reducer

export const selectDocs = (state: RootState): IFetch<IDocs> => state.api.docs
