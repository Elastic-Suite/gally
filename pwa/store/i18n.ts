import { createSlice, PayloadAction } from '@reduxjs/toolkit'

interface II18nState {
  language: string
}

const initialState: II18nState = {
  language: 'en',
}

const i18nSlice = createSlice({
  name: 'i18n',
  initialState,
  reducers: {
    setLanguage(state, action: PayloadAction<string>) {
      state.language = action.payload
    },
  },
})

export const { setLanguage } = i18nSlice.actions
export const i18nReducer = i18nSlice.reducer
