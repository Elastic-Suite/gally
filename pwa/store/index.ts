export * from './api'
export * from './api'
export * from './hooks'
export * from './i18n'
export * from './menu'
export * from './store'
