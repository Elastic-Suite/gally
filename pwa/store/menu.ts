import { createSlice, PayloadAction } from '@reduxjs/toolkit'

export interface MenuChild {
  code: string
  label: string
  children: MenuChild[]
}

export interface Menu {
  hierarchy: MenuChild[]
}

export interface MenuState {
  menu: Menu
  menuItemActive: string
  sidebarState: boolean
  sidebarStateTimeout: boolean
  childrenState: Record<string, boolean>
}

const initialState: MenuState = {
  menu: {
    hierarchy: [],
  },
  menuItemActive: '',
  sidebarState: true,
  sidebarStateTimeout: false,
  childrenState: {},
}

const menuSlice = createSlice({
  name: 'menu',
  initialState,
  reducers: {
    setChildState(
      state,
      action: PayloadAction<{ code: string; value: boolean }>
    ) {
      state.childrenState[action.payload.code] = action.payload.value
    },
    setMenu(state, action: PayloadAction<Menu>) {
      state.menu = action.payload
    },
    setMenuItemActive(state, action: PayloadAction<string>) {
      state.menuItemActive = action.payload
    },
    setSidebarState(state, action: PayloadAction<boolean>) {
      state.sidebarState = action.payload
    },
    setSidebarStateTimeout(state, action: PayloadAction<boolean>) {
      state.sidebarStateTimeout = action.payload
    },
  },
})

export const {
  setChildState,
  setMenu,
  setMenuItemActive,
  setSidebarState,
  setSidebarStateTimeout,
} = menuSlice.actions
export const menuReducer = menuSlice.reducer

export const selectChildrenState = (state, code: string) => state.menu.childrenState[code]
export const selectMenu = (state) => state.menu.menu
export const selectMenuItemActive = (state) => state.menu.menuItemActive
export const selectSidebarState = (state) => state.menu.sidebarState
export const selectSidebarStateTimeout = (state) => state.menu.sidebarStateTimeout
