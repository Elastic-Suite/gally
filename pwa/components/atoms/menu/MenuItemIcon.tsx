import IonIcon from '~/components/atoms/IonIcon'
import { makeStyles } from '@mui/styles'
import { Theme } from '@mui/material/styles'
import Link from 'next/link'
import { useAppSelector } from '~/store'

/*
 * Use of mui makeStyles to create multiple styles reusing theme fm react-admin
 * see: https://mui.com/system/styles/basics/
 */
const useStyles = makeStyles((theme: Theme) => ({
  root: {
    position: 'relative',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: theme.spacing(2),
    textDecoration: 'unset',
    color: 'inherit',
    padding: theme.spacing(1),
    '& ion-icon': {
      color: theme.palette.menu.text500,
    },
  },
  childPadding: {
    paddingBottom: theme.spacing(1),
  },
  boldStyle: {
    fontFamily: 'inter',
    fontWeight: 600,
    fontSize: 13,
    lineHeight: '20px',
    letterSpacing: 0.5,
    textTransform: 'uppercase',
    color: theme.palette.menu.text600,
  },
  lightStyle: {
    fontFamily: 'inter',
    fontWeight: 500,
    fontSize: 14,
    lineHeight: '20px',
    color: theme.palette.menu.text500,
  },
  noChildHover: {
    '&:hover': {
      background: theme.palette.menu.hover,
      borderRadius: theme.spacing(1),
    },
  },
  lineActive: {
    position: 'relative',
    display: 'flex',
    justifyContent: 'space-between',
    color: theme.palette.menu.active,
    flex: 1,
  },
  indicatorLineActive: {
    position: 'absolute',
    right: 0,
    width: 3,
    background: theme.palette.menu.active,
    boxShadow: '-2px 0px 4px rgba(63, 50, 230, 0.2)',
    borderRadius: '5px 0px 0px 5px',
  },
  indicatorLineActiveTwo: {
    width: 3,
    background: theme.palette.menu.active,
    boxShadow: '-2px 0px 4px rgba(63, 50, 230, 0.2)',
    borderRadius: '5px 0px 0px 5px',
  },
  opacityFull: {
    opacity: 0,
    animation: '$opacityFull 1000ms forwards',
  },
  '@keyframes opacityFull': {
    from: { opacity: 0 },
    to: { opacity: 1 },
  },
  opacityFullDeux: {
    opacity: 0,
    animation: '$opacityFullDeux 1200ms forwards',
  },
  '@keyframes opacityFullDeux': {
    '0%': { opacity: 0 },
    '45%': { opacity: 0 },
    '100%': { opacity: 1 },
  },
  span: {
    opacity: 1,
    transition: 'all 500ms',
  },
  hide: {
    opacity: 0,
    height: 0,
  },
  heightZero: {
    opacity: 1,
    maxHeight: 'initial',
    maxWidth: 'initial',
    animation: '$heightZero 1400ms forwards',
    position: 'relative',
  },
  '@keyframes heightZero': {
    '0%': {
      maxHeight: 'initial',
      maxWidth: 'initial',
      opacity: 1,
      position: 'relative',
    },
    '20%': {
      maxHeight: 'initial',
      maxWidth: 'initial',
      opacity: 0,
      position: 'relative',
    },
    '100%': { maxHeight: 0, maxWidth: 0, opacity: 0, position: 'absolute' },
  },
}))

const MenuItemIcon = (props) => {
  const menuItemActive =
    useAppSelector((state) => state.menu.menuItemActive) || ''
  const sidebarState = useAppSelector((state) => state.menu.sidebarState)
  const sidebarStateTimeout = useAppSelector(
    (state) => state.menu.sidebarStateTimeout
  )

  const words = menuItemActive.split('_')
  const wordIndexOne = words[0]

  const classes = useStyles()
  let classNameRoot = classes.root
  let classNameStyle = ''
  if (props.lightStyle) {
    classNameStyle += classes.lightStyle
  } else {
    classNameStyle += classes.boldStyle
  }

  /*
   * Verification and return of specific data if the item has children
   */
  if (props.childPadding) {
    classNameRoot += ' ' + classes.childPadding
    return (
      <div className={classNameRoot + ' ' + classNameStyle}>
        <IonIcon name={props.code} style={{ width: 18, height: 18 }} />
        {sidebarStateTimeout && wordIndexOne === props.code ? (
          <div
            className={
              classes.indicatorLineActive + ' ' + classes.opacityFullDeux
            }
            style={{ height: '32px' }}
          />
        ) : (
          ''
        )}

        <span
          className={
            classes.span +
            (!sidebarStateTimeout
              ? ' ' + classes.opacityFull
              : ' ' + classes.heightZero)
          }
        >
          {props.label}
        </span>
      </div>
    )
  } else {
    return (
      <div
        className={
          menuItemActive === props.code
            ? classes.lineActive + ' ' + classNameStyle
            : classNameStyle
        }
      >
        <Link href="/admin/[[...slug]]" as={`/admin/${props.href}`}>
          <a
            className={classNameRoot + ' ' + classes.noChildHover}
            style={
              sidebarState
                ? { width: 'calc(100% - 10px)' }
                : { width: 'fit-content' }
            }
          >
            <IonIcon name={props.code} style={{ width: 18, height: 18 }} />

            <span
              className={
                classes.span +
                (!sidebarStateTimeout
                  ? ' ' + classes.opacityFull
                  : ' ' + classes.heightZero)
              }
            >
              {props.label}
            </span>
          </a>
        </Link>
        {menuItemActive === props.code && (
          <div
            className={
              classes.indicatorLineActiveTwo +
              (sidebarStateTimeout ? ' ' + classes.opacityFullDeux : '')
            }
          />
        )}
      </div>
    )
  }
}

export default MenuItemIcon
