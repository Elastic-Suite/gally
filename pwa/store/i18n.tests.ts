import { i18nReducer, setLanguage } from './i18n'

describe('i18nReducer', () => {
  it('should return the initial state', () => {
    expect(i18nReducer(undefined, { type: undefined })).toEqual({
      language: 'en',
    })
  })

  it('should set the language', () => {
    const previousState = {
      language: 'en',
    }

    expect(i18nReducer(previousState, setLanguage('fr'))).toEqual({
      language: 'fr',
    })
  })
})
