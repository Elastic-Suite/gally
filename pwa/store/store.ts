import {
  PreloadedState,
  combineReducers,
  configureStore,
} from '@reduxjs/toolkit'
import { apiReducer } from './api'
import { i18nReducer } from './i18n'
import { menuReducer } from './menu'

const rootReducer = combineReducers({
  api: apiReducer,
  i18n: i18nReducer,
  menu: menuReducer,
})

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: {
          ignoredActions: ['api/doc'],
          ignoredActionPaths: ['payload'],
          ignoredPaths: ['api.doc.data'],
        },
      }),
    preloadedState,
    reducer: rootReducer,
  })
}

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']

export interface IThunkApi {
  dispatch: AppDispatch
  state: RootState
}
