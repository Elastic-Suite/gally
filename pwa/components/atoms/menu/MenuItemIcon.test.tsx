import { render } from '@testing-library/react'
import MenuItemIcon from './MenuItemIcon'
import { Provider } from 'react-redux'
import { store } from '~/store'
import { ThemeProvider } from '@mui/material/styles'
import RegularTheme from '~/components/atoms/RegularTheme'

const wrapper = ({ children }) => (
  <Provider store={store}>
    <ThemeProvider theme={RegularTheme}>{children}</ThemeProvider>
  </Provider>
)

describe('MenuItemIcon', () => {
  it('render properly lightStyle with no child (Settings)', () => {
    const { container } = render(
      <MenuItemIcon
        href="settings"
        code="settings"
        label="Settings"
        lightStyle
        childPadding={false}
      />,
      { wrapper }
    )

    expect(container).toBeTruthy()
  })

  it('match snapshot with boldStyle and child (Analyze)', () => {
    const { container } = render(
      <MenuItemIcon
        href="analyze"
        code="analyze"
        label="Analyze"
        lightStyle={false}
        childPadding
      />,
      { wrapper }
    )

    expect(container).toMatchSnapshot()
  })
})
