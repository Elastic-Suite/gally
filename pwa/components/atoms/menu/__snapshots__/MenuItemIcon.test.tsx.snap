// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuItemIcon match snapshot with boldStyle and child (Analyze) 1`] = `
<div>
  <div
    class="css-lalpws"
  >
    <ion-icon
      name="analytics"
      style="width: 18px; height: 18px;"
    />
    
    <span
      class="css-1skr4lk"
    >
      Analyze
    </span>
  </div>
</div>
`;
